### PR TITLE
examples(source-map): useful config

### LINKED ISSUE #9129: Change devtool example to something else
https://github.com/webpack/webpack/blob/master/examples/source-map/webpack.config.js

Your docs point to this this example that is not very useful. That `webpack.config.js` resembles no `webpack.config.js` I've ever encountered. For examples that demonstrate complicated features like devtool, please make an example that resembles a real world project.

### PR DESCRIPTION
**Pull request by shaodahong**
**What kind of change does this PR introduce?**
useful examples source-map
Fixed #9129

**Did you add tests for your changes?**
No

**Does this PR introduce a breaking change?**
No

**What needs to be documented once your changes are merged?**
Yes

### CODE REVIEW

**Comment by shaodahong** [comment]
Fixd #9129

File: examples/source-map/webpack.config.js [hunk 1]
```diff
@@ -1,27 +1,34 @@
 var path = require("path");
 
 module.exports = [
```
Comment (sokra) [inline examples/source-map/webpack.config.js]: You could write the config like that:

``` js
module.exports = (env = "development") => ({
  mode: env,
  devtool: env === "development" ? "cheap-eval-source-map" : "source-map",
  entry: "./example.coffee",
  module: {
    rules: [
      { test: /\.coffee$/, use: "coffee-loader" }
    ]
  }
});
```

`optimization` and `output` could be omitted. This should be a more realistic usage.

File: examples/source-map/template.md [hunk 2]
```diff
@@ -14,72 +14,20 @@ _{{webpack.config.js}}_
...
 
 ```javascript
-_{{dist/bundle-cheap-module-eval-source-map.js}}_
-```
-
-## cheap-module-source-map.js.map
-
-```javascript
-_{{dist/bundle-cheap-module-source-map.js.map}}_
+_{{dist/bundle-source-map.js}}_
```
Comment (sokra) [inline examples/source-map/template.md]:
```suggestion
_{{production:dist/output.js}}_
```

File: examples/source-map/template.md [hunk 3]
```diff
@@ -14,72 +14,20 @@ _{{webpack.config.js}}_
...
 ```
 
-## cheap-source-map.js.map
-
 ```javascript
-_{{dist/bundle-cheap-source-map.js.map}}_
+_{{dist/bundle-source-map.js.map}}_
```
Comment (sokra) [inline examples/source-map/template.md]:
```suggestion
_{{production:dist/output.js.map}}_
```

File: examples/source-map/template.md [hunk 4]
```diff
@@ -14,72 +14,20 @@ _{{webpack.config.js}}_
...
-
-```javascript
-_{{dist/bundle-eval.js}}_
-```
-
-## cheap-eval-source-map.js
+## development mode
 
 ```javascript
 _{{dist/bundle-cheap-eval-source-map.js}}_
```
Comment (sokra) [inline examples/source-map/template.md]:
```suggestion
_{{development:dist/output.js}}_
```

File: examples/source-map/webpack.config.js (hunk 1 above)
Comment (shaodahong) [reply examples/source-map/webpack.config.js]: yep, But I export a function config, and log `env` param, is always `undefined`

```js
module.exports = env => {
	return {
		mode: env,
		entry: "./example.coffee",
		devtool: env === "development" ? "cheap-eval-source-map" : "source-map",
		module: {
			rules: [{ test: /\.coffee$/, use: "coffee-loader" }]
		}
	};
};
```

when I set the default value is `development`, is always `development`, where did I write it wrong?

and I change `build.common.js` add `--env env mode`, it's be ok, but I found development result file `main.js` worng, it's not `cheap-eval-source-map` devtool

```js
async.series([
	callback => doCompileAndReplace("--mode production --env production", "production", callback),
	callback => doCompileAndReplace("--mode development --env development --devtool none", "development", callback),
	callback => doCompileAndReplace("--mode none --output-pathinfo", "", callback)
], () => {
	readme = tc.replaceBase(readme);
	fs.writeFile("README.md", readme, "utf-8", function() {});
});
```

**Comment by shaodahong** [comment]
@sokra I changed `build-common.js`, but I don't know the effect other, help me thanks

File: examples/source-map/webpack.config.js (hunk 1 above)
Comment (shaodahong) [reply examples/source-map/webpack.config.js]: webpack cli `--devtool none` priority higher

**Comment by alexander-akait** [comment]
@shaodahong Sorry for big delay, let's close because it is outdated, anyway if you still want to improve it - feel free to send PR (there will be no more delays, I promise :smile: )

**Pull request closed without merging**